asab/library/service.py

```diff
@@ -12,69 +12,94 @@
 #
 
 
-asab.Config.add_defaults(
-	{
-		'library': {
-			# 'providers'': ''
-		}
-	}
-)
+class LibraryService(Service):
+	'''
+	The library service is designed to "exists" in multiple instances,
+	with different `paths` setup.
+	'''
 
+	def __init__(self, app, service_name, paths=None):
+		super().__init__(app, service_name)
+		self.Libraries = dict()
 
-class LibraryService(asab.Service):
+		if paths is None:
+			try:
+				paths = Config.get("library", "providers")
+			except configparser.NoOptionError:
+				L.critical("'providers' option is not present in configuration section 'library'.")
+				raise SystemExit("Exit due to a critical configuration error.")
 
-	def __init__(self, app, service_name):
-		super().__init__(app, service_name)
-		try:
-			asab.Config.get("library", 'providers')
-		except configparser.NoOptionError:
-			L.critical("'providers' option is not present in configuration section 'library'.")
-			raise SystemExit("Exit due to a critical configuration error.")
+		if isinstance(paths, str):
+			paths = re.split(r"\s+", paths)
 
-		paths = asab.Config["library"]["providers"]
-		self.Libraries = dict()
-		for path in re.split(r"\s+", paths):
+		for path in paths:
 			self._create_library(path)
 
+
+	async def finalize(self, app):
+		while len(self.Libraries) > 0:
+			key = next(iter(self.Libraries))
+			lib = self.Libraries.pop(key)
+			await lib.finalize()
+
+
 	def _create_library(self, path):
 		library_provider = None
 		if path.startswith('zk://') or path.startswith('zookeeeper://'):
+			from .providers.zookeeper import ZooKeeperLibraryProvider
 			library_provider = ZooKeeperLibraryProvider(self.App, path)
 
 		elif path.startswith('./') or path.startswith('/') or path.startswith('file://'):
+			from .providers.filesystem import FileSystemLibraryProvider
 			library_provider = FileSystemLibraryProvider(self.App, path)
 
 		else:
-			L.error("Incorrect providers passed.")
+			L.error("Incorrect/unknow provider for '{}'".format(path))
 			raise SystemExit("Exit due to a critical configuration error.")
 
 		self.Libraries[path] = library_provider
 
-	async def read(self, path):
+
+	# TODO: Read disabled from the first library and apply that on the results of `read` and `list`.
+
+
+	async def read(self, path, tenant=None):
 		for library in self.Libraries.values():
 			item = await library.read(path)
-			if item is not None:
-				return item
+			if item is None:
```

Comment:
@ateska it cannot be None. Unless exception is handled for 'NoNodeError' and then you set the item to None at read method for appropriate providers

Reply:
`NoNodeError` must be captured in the provider ... it is too "Zookeeper-specific" exception to be propagated out of this method.